File: backend/src/api/attendance.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from datetime import date, datetime, timezone
import pytz
from core.dependencies import get_current_admin
from db.database import get_db
from db.models import Attendance, User, UserRole

router = APIRouter(prefix="/attendance", tags=["Attendance"])
IST = pytz.timezone("Asia/Kolkata")

# ------------------ MARK ATTENDANCE (STUDENT) ------------------

@router.post("/mark/{user_id}")
def mark_attendance(
    user_id: str,
    classroom_id: str = Query(...),
    db: Session = Depends(get_db),
):
    # Ensure USER exists in this classroom (NOT admin)
    user = (
        db.query(User)
        .filter(
            User.user_id == user_id,
            User.classroom_id == classroom_id,
            User.role == UserRole.USER
        )
        .first()
    )

    if not user:
        raise HTTPException(404, "User not found in this classroom")

    today = date.today()

    # Prevent duplicate attendance
    exists = (
        db.query(Attendance)
        .filter(
            Attendance.user_id == user_id,
            Attendance.classroom_id == classroom_id,
            Attendance.date == today
        )
        .first()
    )

    if exists:
        return {
            "message": "Attendance already marked for today",
            "user_id": user_id,
            "classroom_id": classroom_id,
        }

    attendance = Attendance(
        user_id=user.user_id,
        classroom_id=classroom_id,
        date=today,
        time=datetime.now(IST).time(),
    )

    db.add(attendance)
    db.commit()

    return {
        "message": "Attendance marked successfully",
        "user_id": user_id,
        "classroom_id": classroom_id,
        "date": today,
    }


# ------------------ READ ATTENDANCE (PUBLIC) ------------------

@router.get("/today")
def get_today_attendance(
    classroom_id: str = Query(...),
    db: Session = Depends(get_db),
):
    today = date.today()

    records = (
        db.query(Attendance, User)
        .join(User, Attendance.user_id == User.user_id)
        .filter(
            Attendance.date == today,
            Attendance.classroom_id == classroom_id,
            User.classroom_id == classroom_id,
            User.role == UserRole.USER
        )
        .all()
    )

    return {
        "date": today,
        "classroom_id": classroom_id,
        "count": len(records),
        "records": [
            {
                "user_id": user.user_id,
                "time": attendance.time.strftime("%H:%M:%S"),
            }
            for attendance, user in records
        ],
    }


@router.get("/by-date")
def get_attendance_by_date(
    classroom_id: str = Query(...),
    attendance_date: date = Query(...),
    db: Session = Depends(get_db),
):
    records = (
        db.query(Attendance, User)
        .join(User, Attendance.user_id == User.user_id)
        .filter(
            Attendance.date == attendance_date,
            Attendance.classroom_id == classroom_id,
            User.classroom_id == classroom_id,
            User.role == UserRole.USER
        )
        .all()
    )

    return {
        "date": attendance_date,
        "classroom_id": classroom_id,
        "count": len(records),
        "records": [
            {
                "user_id": user.user_id,
                "time": attendance.time.strftime("%H:%M:%S"),
            }
            for attendance, user in records
        ],
    }


@router.get("/all")
def get_all_attendance_for_classroom(
    classroom_id: str = Query(...),
    db: Session = Depends(get_db),
):
    records = (
        db.query(Attendance, User)
        .join(User, Attendance.user_id == User.user_id)
        .filter(
            Attendance.classroom_id == classroom_id,
            User.classroom_id == classroom_id,
            User.role == UserRole.USER
        )
        .order_by(Attendance.date.desc(), Attendance.time.asc())
        .all()
    )

    return {
        "classroom_id": classroom_id,
        "count": len(records),
        "records": [
            {
                "user_id": user.user_id,
                "date": attendance.date,
                "time": attendance.time.strftime("%H:%M:%S"),
            }
            for attendance, user in records
        ],
    }


@router.get("/status/{user_id}")
def get_day_wise_attendance_status(
    user_id: str,
    classroom_id: str = Query(...),
    db: Session = Depends(get_db),
):
    user = (
        db.query(User)
        .filter(
            User.user_id == user_id,
            User.classroom_id == classroom_id,
            User.role == UserRole.USER
        )
        .first()
    )

    if not user:
        raise HTTPException(404, "User not found in this classroom")

    all_dates = (
        db.query(Attendance.date)
        .filter(Attendance.classroom_id == classroom_id)
        .distinct()
        .order_by(Attendance.date)
        .all()
    )

    present_dates = (
        db.query(Attendance.date)
        .filter(
            Attendance.user_id == user_id,
            Attendance.classroom_id == classroom_id
        )
        .all()
    )

    present_set = {d[0] for d in present_dates}

    records = [
        {
            "date": att_date,
            "status": "PRESENT" if att_date in present_set else "ABSENT",
        }
        for (att_date,) in all_dates
    ]

    return {
        "user_id": user_id,
        "classroom_id": classroom_id,
        "total_days": len(all_dates),
        "present_days": len(present_set),
        "absent_days": len(all_dates) - len(present_set),
        "records": records,
    }


# ------------------ ADMIN DELETE ENDPOINTS ------------------

@router.delete("/admin/today/{user_id}")
def delete_today_attendance_for_user(
    user_id: str,
    classroom_id: str = Query(...),
    db: Session = Depends(get_db),
    admin = Depends(get_current_admin),
):
    today = date.today()

    record = (
        db.query(Attendance)
        .filter(
            Attendance.user_id == user_id,
            Attendance.classroom_id == classroom_id,
            Attendance.date == today
        )
        .first()
    )

    if not record:
        raise HTTPException(404, "Attendance record not found")

    db.delete(record)
    db.commit()

    return {
        "message": "Today's attendance deleted",
        "user_id": user_id,
        "classroom_id": classroom_id,
    }


@router.delete("/admin/today")
def delete_all_today_attendance(
    classroom_id: str = Query(...),
    db: Session = Depends(get_db),
    admin = Depends(get_current_admin),
):
    today = date.today()

    records = (
        db.query(Attendance)
        .filter(
            Attendance.date == today,
            Attendance.classroom_id == classroom_id
        )
        .all()
    )

    for record in records:
        db.delete(record)

    db.commit()

    return {
        "message": "All today's attendance deleted",
        "classroom_id": classroom_id,
        "count": len(records),
    }


@router.delete("/admin/user/{user_id}")
def delete_all_attendance_for_user(
    user_id: str,
    classroom_id: str = Query(...),
    db: Session = Depends(get_db),
    admin = Depends(get_current_admin),
):
    records = (
        db.query(Attendance)
        .filter(
            Attendance.user_id == user_id,
            Attendance.classroom_id == classroom_id
        )
        .all()
    )

    if not records:
        raise HTTPException(404, "No attendance records found")

    for record in records:
        db.delete(record)

    db.commit()

    return {
        "message": "All attendance records deleted for user",
        "user_id": user_id,
        "count": len(records),
    }

File: frontend/src/api/attendance.js
// frontend/src/api/attendance.js

/**
 * Frontend API layer for:
 * backend/src/api/attendance.py
 */

const API_BASE = "http://localhost:8002";

/* ======================================================
   MARK ATTENDANCE
   ====================================================== */

/**
 * Mark attendance for a user in a classroom
 *
 * Backend:
 * POST /attendance/mark/{user_id}?classroom_id=
 */
export async function markAttendance(userId, classroomId) {
  const response = await fetch(
    `${API_BASE}/attendance/mark/${userId}?classroom_id=${classroomId}`,
    {
      method: "POST",
    }
  );

  if (!response.ok) {
    throw new Error(await response.text());
  }

  return response.json();
}

/* ======================================================
   GET ATTENDANCE (CLASSROOM)
   ====================================================== */

/**
 * Get today's attendance for a classroom
 *
 * Backend:
 * GET /attendance/today?classroom_id=
 */
export async function fetchTodayAttendance(classroomId) {
  const response = await fetch(
    `${API_BASE}/attendance/today?classroom_id=${classroomId}`
  );

  if (!response.ok) {
    throw new Error(await response.text());
  }

  return response.json();
}

/**
 * Get attendance for a classroom by date
 *
 * Backend:
 * GET /attendance/by-date?classroom_id=&attendance_date=YYYY-MM-DD
 */
export async function fetchAttendanceByDate(classroomId, attendanceDate) {
  const response = await fetch(
    `${API_BASE}/attendance/by-date?classroom_id=${classroomId}&attendance_date=${attendanceDate}`
  );

  if (!response.ok) {
    throw new Error(await response.text());
  }

  return response.json();
}

/**
 * Get all attendance records for a classroom (all dates)
 *
 * Backend:
 * GET /attendance/all?classroom_id=
 */
export async function fetchAllAttendanceForClassroom(classroomId) {
  const response = await fetch(
    `${API_BASE}/attendance/all?classroom_id=${classroomId}`
  );

  if (!response.ok) {
    throw new Error(await response.text());
  }

  return response.json();
}

/* ======================================================
   USER ATTENDANCE STATUS
   ====================================================== */

/**
 * Get day-wise attendance status for a user in a classroom
 *
 * Backend:
 * GET /attendance/status/{user_id}?classroom_id=
 */
export async function fetchUserAttendanceStatus(userId, classroomId) {
  const response = await fetch(
    `${API_BASE}/attendance/status/${userId}?classroom_id=${classroomId}`
  );

  if (!response.ok) {
    throw new Error(await response.text());
  }

  return response.json();
}

/* ======================================================
   DELETE ATTENDANCE
   ====================================================== */

/**
 * Delete today's attendance for a user (classroom-specific)
 *
 * Backend:
 * DELETE /attendance/today/{user_id}?classroom_id=
 */
export async function deleteTodayAttendanceForUser(userId, classroomId) {
  const response = await fetch(
    `${API_BASE}/attendance/today/${userId}?classroom_id=${classroomId}`,
    {
      method: "DELETE",
    }
  );

  if (!response.ok) {
    throw new Error(await response.text());
  }

  return response.json();
}

/**
 * Delete all today's attendance for a classroom
 *
 * Backend:
 * DELETE /attendance/today?classroom_id=
 */
export async function deleteAllTodayAttendance(classroomId) {
  const response = await fetch(
    `${API_BASE}/attendance/today?classroom_id=${classroomId}`,
    {
      method: "DELETE",
    }
  );

  if (!response.ok) {
    throw new Error(await response.text());
  }

  return response.json();
}

/**
 * Delete all attendance records for a user in a classroom
 *
 * Backend:
 * DELETE /attendance/user/{user_id}?classroom_id=
 */
export async function deleteAllAttendanceForUser(userId, classroomId) {
  const response = await fetch(
    `${API_BASE}/attendance/user/${userId}?classroom_id=${classroomId}`,
    {
      method: "DELETE",
    }
  );

  if (!response.ok) {
    throw new Error(await response.text());
  }

  return response.json();
}
